src/kbmod/search/logging.h

```diff
@@ -0,0 +1,221 @@
+    void error(std::string msg) { log("ERROR", msg); }
+    void critical(std::string msg) { log("CRITICAL", msg); }
+  };
+
+
+#ifdef Py_PYTHON_H
+  // Passes the logging to the Python-side Logger. The py::handle is the
+  // pybind11 container holding the reference to the Python Logger. The actual
+  // logging format and mechanism is handled Python-side by the logging module,
+  // so just pass the message and level onward and let it be resolved in Python
+  class PyLogger : public Logger {
+  private:
+    py::handle pylogger;
+
+  public:
+    PyLogger(py::handle logger) :
+      Logger(logger.attr("name").cast<std::string>()),
+      pylogger(logger)
+    {}
+
+    virtual void log(std::string level, std::string msg){
+      for (char& ch : level) ch = std::tolower(ch);
+      pylogger.attr(level.c_str())(msg);
+    }
+  };
+#endif // Py_PYTHON_H
+
+
+  // Glorified std::cout. Unlike the PyLogger, CoutLogger is configurable.
+  // Configuration should contain `level`, `datefmt` and `format` keys:
+  // - `level`:  LogLevel enum value, the minimal level that is printed
+  // - `datefmt`: timestamp template usable with `strftime`
+  // - `format`: log format template, too much commitment atm, not supported
+  class CoutLogger : public Logger{
+  private:
+    // datefmt %Y-%m expands to YYYY-MM etc., 2x its size sounds safe enough but
+    // with abuse I guess this could overflow?
+    std::string fmt_time(){
+      time_t now;
+      time(&now);
+      char buf[2*sizeof(config["datefmt"])] = {0};
+      strftime(buf, sizeof buf, config["datefmt"].c_str(), gmtime(&now));
+      return std::string(buf);
+    }
+
+  public:
+    CoutLogger(std::string name, sdict config) :
+      Logger(name, config)
+    {}
+
+    virtual void log(std::string level, std::string msg){
+      if (level_threshold <= StringToLogLevel[level])
+        std::cout << "[" << fmt_time() << " " << level << " " << name << "] " << msg << std::endl;
+    }
+  };
+
+
+  // The singleton keeping the registry of all registered Loggers and their
+  // default configuration. Use `logger` to access the singleton instance,
+  // getLoger to get an existing logger or to create a new default logger.
+  // No good examples, but let's say a non-stdout C++ only logger is required,
+  // use getLogger<LoggerCls>(name, conf) to register and get the instance of it
+  class Logging{
+  private:
+    static sdict default_config;
+    std::map<std::string, Logger*> registry;
+    static Logging* instance;
+
+    Logging(){}
+    ~Logging(){}
+
+  public:
+    Logging(Logging &other) = delete;
+    void operator=(const Logging &) = delete;
+
+    // get the singleton instance
+    static Logging* logger(){
+      if(instance == nullptr)
+        instance = new Logging();
+      return instance;
+    }
+
+    static void setConfig(sdict config){
+      default_config = config;
+    }
+
+    template<class LoggerCls>
+    static Logger* getLogger(std::string name, sdict config={}){
```

Comment:
What are the use cases for a per-logger config? I would imagine most of the time we'd be okay with a single global config.

Reply:
It's how Python does it, so I just cargo-culted. The option is there, but it's currently not being used. They are all just inheriting a central global config and using that.